### PR TITLE
Select enhancement

### PR DESCRIPTION
**Pull request by mmakaay**
# What does this implement/fix?

This PR adds a couple of useful automation actions to the `select` component, for selecting the first, last, next or previous item in the select component. This  resembles the [automation functionalities that are available for Home Assistant](https://www.home-assistant.io/integrations/input_select/#services).
These actions make use cases like "select an option from the select list using a dial knob" trivial.

Next to the explicit actions ``select.next``, ``select.previous``, ``select.first`` and ``select.last``, there is also a generic ``select.operation`` action, that takes a templatable ``operation`` parameter, to make it possible to perform a dynamic operation based on a lambda.

To make it easier to work with index-based access, various methods have been added to the select component API (``index_of(<value>)``, ``at(<index>)``, ``active_index()``, ``size()``). Also, an extra parameter ``i`` was added to the even trigger so event handlers can use the index offset directly, without having to perform a lookup.

An automation action `select.set_index` was added to complement the above reader methods. Using this method, one can set the selected option based on the index offset, instead of the option value.

Additionally, I split up the various classes off to their own .h and .cpp files.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2046

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
select:
  - platform: template
    name: "My list of things"
    id: thing_from_list
    optimistic: true
    options:
      - apple
      - bike
      - candy
    on_value:
      then:
        logger.log:
          format: "Option selected: %s (index %d)"
          args: ["x.c_str()", "i"]

sensor:
  platform: rotary_encoder
  on_clockwise:
    then:
      select.next: thing_from_list
  on_anticlockwise:
    then:
      select.previous:
        id: thing_from_list
        cycle: false
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by mmakaay** [comment]
In the original code, I found two bits that aren't documented, nor used.
Question is: what to do with those bits. The parts that I'm talking about are:

**Method ``Select::set(...)``**
``void set(const std::string &value) { make_call().set_option(value).perform(); }``

In the documentation, `make_call()` is promoted for making state changes. If this one is here to stay, then I would find it logical to implement ``next(<cycle?>)``, ``first()``, etc. as direct methods on the Select class as well, and include these shortcut utility methods in the documentation. If not, then let's remove it.

**Method ``SelectCall::get_option()``**
``const optional<std::string> &SelectCall::get_option() const { return option_; }``

This method doesn't feel that useful on the ``SelectCall`` class. The only real use that I can see, would be to have some method of inspection when writing unit tests. 

Do you have some input on these two?

File: esphome/components/select/__init__.py [hunk 1]
```diff
@@ -22,14 +26,27 @@
 
 # Triggers
 SelectStateTrigger = select_ns.class_(
-    "SelectStateTrigger", automation.Trigger.template(cg.float_)
```
Comment (jesserockz) [inline esphome/components/select/__init__.py:25]: Woops, my bad

**Review by jesserockz — APPROVED** [review]
Looks good to me

**Comment by jesserockz** [comment]
> In the original code, I found two bits that aren't documented, nor used. Question is: what to do with those bits. The parts that I'm talking about are:
> 
> **Method `Select::set(...)`** `void set(const std::string &value) { make_call().set_option(value).perform(); }`
> 
> In the documentation, `make_call()` is promoted for making state changes. If this one is here to stay, then I would find it logical to implement `next(<cycle?>)`, `first()`, etc. as direct methods on the Select class as well, and include these shortcut utility methods in the documentation. If not, then let's remove it.
> 
> **Method `SelectCall::get_option()`** `const optional<std::string> &SelectCall::get_option() const { return option_; }`
> 
> This method doesn't feel that useful on the `SelectCall` class. The only real use that I can see, would be to have some method of inspection when writing unit tests.
> 
> Do you have some input on these two?

Happy to get rid of them (In a followup PR :smile:)

**Pull request merged**